### PR TITLE
refactor and clean code

### PR DESCRIPTION
**Pull request by georgkcode0**
- Check installed package in mac os
- Use `template` for service configuration
- Change `rpc-port` from `string` to `int`
- Add `onnxruntime` dependency for dd-service
- Use go-ps package for check running process
- Add some support files for dd service
- Clean code

### CODE REVIEW

**Review by a-ok123 — DISMISSED** [review]

File: cmd/install.go [hunk 1]
```diff
@@ -317,7 +317,13 @@ func runComponentsInstall(ctx context.Context, config *configs.Config, installCo
 		installCommand == constants.SuperNode {
 
 		toolPath := constants.PastelRQServiceExecName[utils.GetOS()]
-		toolConfig := fmt.Sprintf(configs.RQServiceConfig, "127.0.0.1", "50051")
+		toolConfig, err := utils.GetServiceConfig("rqservice", configs.RQServiceDefaultConfig, &configs.RQServiceConfig{
+			HostName: "127.0.0.1",
+			Port:     50051,
+		})
+		if err != nil {
+			return errors.Errorf("failed to get service config, err: %v", err)
```
Comment (pjenys0) [inline cmd/install.go:325]: nit: should change to errors.Errorf("failed to get rqservice config: %v", err)

File: cmd/install.go [hunk 2]
```diff
@@ -335,9 +341,16 @@ func runComponentsInstall(ctx context.Context, config *configs.Config, installCo
...
 		toolPath := constants.WalletNodeExecName[utils.GetOS()]
-		toolConfig := fmt.Sprintf(configs.WalletDefaultConfig, config.RPCPort, config.RPCUser, config.RPCPwd, "50051")
+		toolConfig, err := utils.GetServiceConfig("walletnode", configs.WalletDefaultConfig, &configs.WalletNodeConfig{
+			PastelPort:     config.RPCPort,
+			PastelUserName: config.RPCUser,
+			PastelPassword: config.RPCPwd,
+			RaptorqPort:    50051,
+		})
+		if err != nil {
+			return errors.Errorf("failed to get service config, err: %v", err)
```
Comment (pjenys0) [inline cmd/install.go:352]: nit: should change to errors.Errorf("failed to get walletnode config: %v", err)

File: cmd/install.go [hunk 3]
```diff
@@ -354,9 +367,16 @@ func runComponentsInstall(ctx context.Context, config *configs.Config, installCo
...
 		toolPath := constants.SuperNodeExecName[utils.GetOS()]
-		toolConfig := fmt.Sprintf(configs.SupernodeDefaultConfig, config.RPCPort, config.RPCUser, config.RPCPwd, "50051")
+		toolConfig, err := utils.GetServiceConfig("supernode", configs.SupernodeDefaultConfig, &configs.SuperNodeConfig{
+			PastelPort:     config.RPCPort,
+			PastelUserName: config.RPCUser,
+			PastelPassword: config.RPCPwd,
+			RaptorqPort:    50051,
+		})
+		if err != nil {
+			return errors.Errorf("failed to get service config, err: %v", err)
```
Comment (pjenys0) [inline cmd/install.go:378]: nit: should change to errors.Errorf("failed to get supernode config: %v", err)

File: cmd/install.go [hunk 4]
```diff
@@ -715,8 +733,8 @@ func installDupeDetection(ctx context.Context, config *configs.Config) (err erro
 
 	ddBaseDir := filepath.Join(config.Configurer.GetHomeDir(), "pastel_dupe_detection_service")
 	var pathList []interface{}
-	for index := range constants.DupeDetectionConfigs {
-		dupeDetectionDirPath := filepath.Join(ddBaseDir, constants.DupeDetectionConfigs[index])
+	for _, name := range constants.DupeDetectionConfigs {
```
Comment (pjenys0) [inline cmd/install.go:736]: nit: name -> configItem

File: cmd/start.go [hunk 5]
```diff
@@ -288,8 +287,22 @@ func runMasterNodOnHotHot(ctx context.Context, config *configs.Config) error {
 		return err
 	}
 
+	pastelPort, err := strconv.Atoi(pastelID)
```
Comment (pjenys0) [inline cmd/start.go:290]: why Port is determined from pastelID???

File: cmd/start.go [hunk 6]
```diff
@@ -288,8 +287,22 @@ func runMasterNodOnHotHot(ctx context.Context, config *configs.Config) error {
...
+		return errors.Errorf("failed to convert %s to integer, err: %v", pastelID, err)
+	}
+	toolConfig, err := utils.GetServiceConfig("supernode", configs.SupernodeDefaultConfig, &configs.SuperNodeConfig{
+		PastelPort:     pastelPort,
+		PastelUserName: extIP,
+		PastelPassword: extPort,
+		RaptorqPort:    50051,
+	})
+	if err != nil {
+		return errors.Errorf("failed to get service config, err: %v", err)
```
Comment (pjenys0) [inline cmd/start.go:301]: nit: failed to get supernode config

File: cmd/start.go (hunk 5 above)
Comment (georgkcode0) [reply cmd/start.go:290]: just follow current code. we need to verify again

File: cmd/stop.go [hunk 7]
```diff
@@ -227,53 +226,25 @@ func stopPatelCLI(ctx context.Context, config *configs.Config) (output string, e
...
+	for _, p := range proc {
+		if p.Executable() == execName {
+			pid = p.Pid()
+			break
 		}
+	}
 
+	process, err := os.FindProcess(pid)
+	if err != nil {
+		return errors.Errorf("failed to find process, err: %v", err)
```
Comment (pjenys0) [inline cmd/stop.go:245]: should include process name in error mesage

File: utils/utils.go [hunk 8]
```diff
@@ -394,22 +397,53 @@ func GetChecksum(_ context.Context, fileName string) (checksum string, err error
...
+			files, err := ioutil.ReadDir(path)
+			if err != nil {
+				log.WithContext(ctx).Errorf("failed to read dir, err: %s", err)
+				continue
+			}
+
+			for _, file := range files {
+				m[file.Name()] = true
+			}
 		}
```
Comment (pjenys0) [inline utils/utils.go:431]: what about other OS - like windows?

Comment (georgkcode0) [reply utils/utils.go:431]: current not implement yet. i don't familiar with windows and will investigate later.

File: utils/utils.go [hunk 9]
```diff
@@ -394,22 +397,53 @@ func GetChecksum(_ context.Context, fileName string) (checksum string, err error
...
 	}
 	return m
 }
+
+// GetServiceConfig returns service configuration
+func GetServiceConfig(name string, format string, value interface{}) (string, error) {
+	temp, err := template.New(name).Parse(format)
```
Comment (leonidcode20) [inline utils/utils.go:438]: Is this `html/template` or `texte/template` ?

File: cmd/stop.go [hunk 10]
```diff
@@ -227,53 +226,25 @@ func stopPatelCLI(ctx context.Context, config *configs.Config) (output string, e
...
+	for _, p := range proc {
+		if p.Executable() == execName {
+			pid = p.Pid()
+			break
 		}
+	}
 
+	process, err := os.FindProcess(pid)
+	if err != nil {
+		return errors.Errorf("failed to find process, process: %s, err: %v", execName, err)
```
Comment (pjenys0) [inline cmd/stop.go:245]: don't need put `err:`

File: cmd/stop.go [hunk 11]
```diff
@@ -227,53 +226,25 @@ func stopPatelCLI(ctx context.Context, config *configs.Config) (output string, e
...
-					proc.Kill()
-
-					break
-				}
-			}
+func processKill(toolType constants.ToolType) error {
+	execName := constants.ServiceName[toolType][utils.GetOS()]
+	proc, err := ps.Processes()
+	if err != nil {
+		return errors.Errorf("failed to get list process, err: %v", err)
```
Comment (pjenys0) [inline cmd/stop.go:233]: same, dont need to add "err:" in error message

File: utils/utils.go (hunk 9 above)
Comment (georgkcode0) [reply utils/utils.go:438]: ahh. should `text/template`. will fix in next commit

**Review by pjenys0 — DISMISSED** [review]
LGTM.
Don't forget to support windows

File: utils/utils.go (hunk 9 above)
Comment (georgkcode0) [reply utils/utils.go:438]: i think we should use `text/template`

File: cmd/start.go (hunk 5 above)
Comment (a-ok123) [reply cmd/start.go:290]: oh, i am just going over start.go logic now - this is total shit show

**Review by a-ok123 — DISMISSED** [review]

**Review by a-ok123 — APPROVED** [review]

**Pull request merged**